precompiles/ArbAggregator.go

```diff
@@ -20,87 +16,111 @@ type ArbAggregator struct {
 	Address addr // 0x6d
 }
 
-// Gets an account's preferred aggregator
+var ErrNotOwner = errors.New("must be called by chain owner")
+
+// [Deprecated]
 func (con ArbAggregator) GetPreferredAggregator(c ctx, evm mech, address addr) (prefAgg addr, isDefault bool, err error) {
-	l1p := c.State.L1PricingState()
-	maybePrefAgg, err := l1p.UserSpecifiedAggregator(address)
+	posters, err := c.State.L1PricingState().BatchPosterTable().AllPosters()
 	if err != nil {
 		return common.Address{}, false, err
 	}
-	if maybePrefAgg != nil {
-		return *maybePrefAgg, false, nil
-	}
-	maybeReimbursableAgg, err := l1p.ReimbursableAggregatorForSender(address)
-	if err != nil || maybeReimbursableAgg == nil {
-		return common.Address{}, false, err
+	if len(posters) == 0 {
+		return common.Address{}, false, errors.New("no batch posters exist")
 	}
-	return *maybeReimbursableAgg, true, nil
+	return posters[0], true, err
 }
 
-// Sets the caller's preferred aggregator to that provided
-func (con ArbAggregator) SetPreferredAggregator(c ctx, evm mech, prefAgg addr) error {
-	var maybePrefAgg *common.Address
-	if prefAgg != (common.Address{}) {
-		maybePrefAgg = &prefAgg
+// [Deprecated]
+func (con ArbAggregator) GetDefaultAggregator(c ctx, evm mech) (addr, error) {
+	posters, err := c.State.L1PricingState().BatchPosterTable().AllPosters()
+	if err != nil {
+		return common.Address{}, err
+	}
+	if len(posters) == 0 {
+		return common.Address{}, errors.New("no batch posters exist")
 	}
-	return c.State.L1PricingState().SetUserSpecifiedAggregator(c.caller, maybePrefAgg)
+	return posters[0], err
 }
 
-// Gets the chain's default aggregator
-func (con ArbAggregator) GetDefaultAggregator(c ctx, evm mech) (addr, error) {
-	return c.State.L1PricingState().DefaultAggregator()
+// Get the addresses of all current batch posters
+func (con ArbAggregator) GetBatchPosters(c ctx, evm mech) ([]addr, error) {
+	return c.State.L1PricingState().BatchPosterTable().AllPosters()
```

Comment:
We do charge for storage, but we don't charge for memory in an exponential manner like Solidity does. We should confirm that our current charging is good enough to allow an unbounded lookup like this.

Reply:
An alternative here would be to get the number of posters, and then manually burn an exponential amount of gas corresponding to how much it'd cost to store that many words in the EVM